curator-recipes/src/main/java/org/apache/curator/framework/recipes/cache/CuratorCacheImpl.java

```diff
@@ -59,14 +59,29 @@ class CuratorCacheImpl implements CuratorCache, CuratorCacheBridge {
     private final StandardListenerManager<CuratorCacheListener> listenerManager = StandardListenerManager.standard();
     private final Consumer<Exception> exceptionHandler;
 
-    private final Phaser outstandingOps = new Phaser() {
+    private final Phaser rootPhaser = new Phaser() {
         @Override
         protected boolean onAdvance(int phase, int registeredParties) {
             callListeners(CuratorCacheListener::initialized);
+            synchronized (CuratorCacheImpl.this) {
+                currentChildPhaser = rootPhaser; 
+            } 
             return true;
         }
     };
 
+    private Phaser currentChildPhaser = new Phaser(rootPhaser);
+
+    private synchronized Phaser getPhaserAndRegister() {
+        if (currentChildPhaser.getRegisteredParties() >= 0xffff) {
+            currentChildPhaser = new Phaser(rootPhaser);
```

Comment:
```suggestion
            currentChildPhaser = new Phaser(currentChildPhaser);
```

?

Reply:
Is there a reason to prefer chain of child phasers versus a single layer deep tree off the single parent root phaser?

I suppose a tree has a limit of 64k children, but that should support 172x10^9 ZNodes.

Reply:
> prefer chain of child phasers versus a single layer deep tree off the single parent root phaser

You are right that the returned currentChildPhaser are referred in the method and called `arriveAndDeregister` accordingly. Then I don't have issue for this implementation.